### ISSUE TITLE
看了代码发现样本组织形式是（q,D+)(q,D-),....而不是论文中的(q,D+,D-,D-,D-，D-),这两种有啥区别呢？

### ISSUE THREAD

**Issue by tianrenheyi09** [shenweichen/DeepMatch#8]
Please refer to the [FAQ](https://deepmatch.readthedocs.io/en/latest/FAQ.html) in doc and search for the [related issues](https://github.com/shenweichen/DeepCTR/issues) before you ask the question.

**Describe the question(问题描述)**
A clear and concise description of what the question is.

**Additional context**

**Operating environment(运行环境):**
 - python version [e.g. 3.6]
 - tensorflow version [e.g. 1.4.0,]
 - deepmatch version [e.g. 0.1.1,]

**Comment by shenweichen** [comment]
这个是将类别数目较大的多分类问题通过负采样转化为一个二分类问题的处理方式。

**Comment by LmYjQ** [comment]
> 这个是将类别数目较大的多分类问题通过负采样转化为一个二分类问题的处理方式。

请问如果是二分类的话，为什么还可以用sampledsoftmaxloss呢？

**Comment by JWenBin** [comment]
> 
> 
> > 这个是将类别数目较大的多分类问题通过负采样转化为一个二分类问题的处理方式。
> 
> 请问如果是二分类的话，为什么还可以用sampledsoftmaxloss呢？

来自百度：

 softmax层的优化：

      这里因为原始语料库通常是上百万甚至更多，如果计算每一个单词为中心词上下文的概率，并反向传播更新每一条链路权重，计算量太大太大。所以这里考虑的是采用负采样或者层次话softmax来计算速度。通常负采样用的比较多，主要是将多分类问题转化为二分类问题，只有正类和负类，其中正类是中心词的上下文单词，负类不属于上下文单词。这里要做的就是最大化正类，最小化负类。

    在负采样中，1个正样本对应设定的K个负样本，所以模型一共由K+1个分类，对比之前将所有语料库中的单词进行softmax分类，负采样转化为了K+1个二分类问题，每次迭代训练的K+1个而不是上百万个，所以权重矩阵也只会更新一小部分，大大提高了计算速度。

   而至于如何进行负采样，有现成的公式，自行百度即可。

**Comment by bbruceyuan** [comment]
> > > 这个是将类别数目较大的多分类问题通过负采样转化为一个二分类问题的处理方式。
> > 
> > 
> > 请问如果是二分类的话，为什么还可以用sampledsoftmaxloss呢？
> 
> 来自百度：
> 
> softmax层的优化：
> 
> ```
>   这里因为原始语料库通常是上百万甚至更多，如果计算每一个单词为中心词上下文的概率，并反向传播更新每一条链路权重，计算量太大太大。所以这里考虑的是采用负采样或者层次话softmax来计算速度。通常负采样用的比较多，主要是将多分类问题转化为二分类问题，只有正类和负类，其中正类是中心词的上下文单词，负类不属于上下文单词。这里要做的就是最大化正类，最小化负类。
> 
> 在负采样中，1个正样本对应设定的K个负样本，所以模型一共由K+1个分类，对比之前将所有语料库中的单词进行softmax分类，负采样转化为了K+1个二分类问题，每次迭代训练的K+1个而不是上百万个，所以权重矩阵也只会更新一小部分，大大提高了计算速度。
> ```
> 
> 而至于如何进行负采样，有现成的公式，自行百度即可。

非常感谢您的回复，这个回复让我收获颇多，但是我觉得您这个回答应该没有解决 @LmYjQ 的问题。

1. 按照我的理解，你解释的是为什么可以用 k + 1个二分类来代替 全量 softmax
2. 按照 tf sample softmax 的实现，需要输入 labels, 而这个 label 应该是 num_labels 对应的 index。如果是二分类的话，那么label只有 0/1， 那么是怎么对应的  num_classes 的呢？tf samplesoftmax 可以参见：https://www.tensorflow.org/api_docs/python/tf/nn/sampled_softmax_loss和解读https://zhuanlan.zhihu.com/p/141421400

期待作者解答一下，感激各位~

**Comment by LmYjQ** [comment]
@bbruceyuan 我又看了下代码，发现DSSM和youtubeDNN构造样本的方式是不一样的。DSSM是在预处理的时候构造了负样本，而youtubeDNN预处理没有特殊处理，在模型里面用了SampledSoftmaxLayer
![Uploading image.png…]()

**Comment by johndkl** [comment]
改成二分类是不是就是纯pointwise啦？ 是不是和DSSM原文 softmax后只计算正样本loss有差别，原文偏pairwise？
